File: 20-valid-parentheses/20-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const stack = [];
  const open = ["(", "[", "{"];
  const object = {
    ")": "(",
    "}": "{",
    "]": "[",
  }
  for (let item of s) {
    if (open.includes(item)) {
      stack.push(item);
    }
    else {
      if (stack[stack.length - 1] === object[item]) {
        stack.pop();
      } else return false;
    }
  }

  return stack.length === 0
}

File: 20-valid-parentheses/20-valid-parentheses.go
func isValid(s string) bool {
	stack := []string{}
	object := map[string]string{
		")": "(",
		"}": "{",
		"]": "[",
	}

	for i := 0; i < len(s); i++ {
		switch string(s[i]) {
		case "(", "[", "{":
			stack = append(stack, string(s[i]))
		default:
			if len(stack) == 0 || stack[len(stack)-1] != object[string(s[i])] {
				return false
			} else {
				stack = stack[:len(stack)-1]
			}
		}
	}
	return len(stack) == 0
}
